In [18]:
import json
import os
import sys
import re

In [3]:
products = []
with open('products.txt') as f:
    lines = f.readlines()
    for line in lines:
        products.append(json.loads(line))

listings = []
with open('listings.txt') as f:
    lines = f.readlines()
    for line in lines:
        listings.append(json.loads(line))

In [4]:
len(products),len(listings)

(743, 20196)

In [15]:
products[0], listings[0]

({'announced-date': '2010-01-06T19:00:00.000-05:00',
  'family': 'Cyber-shot',
  'manufacturer': 'Sony',
  'model': 'DSC-W310',
  'product_name': 'Sony_Cyber-shot_DSC-W310'},
 {'currency': 'CAD',
  'manufacturer': 'Neewer Electronics Accessories',
  'price': '35.99',
  'title': 'LED Flash Macro Ring Light (48 X LED) with 6 Adapter Rings for For Canon/Sony/Nikon/Sigma Lenses'})

In [16]:
type(products[0])

dict

In [6]:
manu = set()
for prod in products:
    manu.add(prod['manufacturer'].lower())

In [7]:
manu

{'agfa',
 'canon',
 'casio',
 'contax',
 'epson',
 'fujifilm',
 'hp',
 'kodak',
 'konica minolta',
 'kyocera',
 'leica',
 'nikon',
 'olympus',
 'panasonic',
 'pentax',
 'ricoh',
 'samsung',
 'sanyo',
 'sigma',
 'sony',
 'toshiba'}

In [15]:
extra_manu = set()
extra = 0
for listing in listings:
    if listing['manufacturer'].lower() not in manu:
        extra += 1
        extra_manu.add(listing['manufacturer'].lower())

In [17]:
extra, extra_manu

(5365,
 {'',
  '"sakar international, inc."',
  '----------------------------',
  '17472',
  '300k',
  '3dinlife',
  '8187049',
  'accessory genie',
  'acer',
  'acer computer gmbh',
  'acmaxx',
  'advance',
  'advent',
  'agfaphoto',
  'aiptek international gmbh',
  'aiptek, inc.',
  'airis',
  'akita electronics',
  'akor',
  'amewi',
  'ansmann',
  'antquarium',
  'apc',
  'apex',
  'apple computer',
  'aqualung',
  'architect range',
  'arctic silver',
  'argus',
  'ativa',
  'avon',
  'axdia international gmbh',
  'axis communication  inc.',
  'b-tech',
  'b.i.g.',
  'baby dan',
  'banshee bungee',
  'barbie',
  'bargainsources',
  'bell & howell',
  'benq',
  'benq deutschland gmbh',
  'bilora',
  'biostek',
  'bosch',
  'bratz',
  'braun photo technik',
  'brookstone',
  'budget hd camcorder',
  'bulova',
  'bushnell',
  'cam',
  'camera',
  'canal toys',
  'canon cameras us',
  'canon canada',
  'canon uk ltd',
  'canon usa',
  'canoon',
  'carat',
  'carat electronics',
  'cas

In [10]:
# check the first word
first_word_not_manuf_count = 0
first_word_not_manuf = dict()
for listing in listings:
    first_word = listing['title'].find(' ')
    if listing['title'][:first_word].lower() not in manu:
        first_word_not_manuf_count += 1
        if listing['title'][:first_word].lower() in first_word_not_manuf:
            first_word_not_manuf[listing['title'][:first_word].lower()] += 1
        else:
            first_word_not_manuf[listing['title'][:first_word].lower()] = 1

In [11]:
first_word_not_manuf_count

3743

In [12]:
len(first_word_not_manuf)

537

In [42]:
print(first_word_not_manuf)

{'dora': 1, 'finepix': 10, 'fe-25silber,': 1, 'samsonite': 10, 'watchguard': 8, 'st5500': 1, 'a835': 1, 'mustek': 13, 'fuj16010760': 1, 'new': 33, 'professional': 3, 'concord': 11, 'ubisoft': 2, 'metabo': 2, 'apn': 1, 'flip': 1, 'osi': 2, 'bundle': 10, 'fisher': 5, 'fototasche,': 105, 'slide': 1, 'd300s+tamron': 2, 'powershot': 8, 'lexar': 1, 'minox': 41, 'm1031-w': 1, 'x-vision': 16, 'insignia': 8, 'minolta': 36, 'sun': 1, '227593:': 1, '22': 3, 'dmc-fs35': 3, 'svp\xa0aqua\xa0hdv1000\xa0(mit\xa0micro8gb)\xa0red\xa0wasserdicht\xa0camcorde': 1, 'new\xa0wp5300\xa0rose\xa0(avec\xa0micro\xa0de': 1, 'bundle:': 2, '32gb': 13, 'cullmann': 2, 'svp\xa0aqua\xa0blue\xa0hdv1000\xa0wasserdicht\xa0camcorde': 1, 'i1437': 2, 'imo': 1, 'camera,digital,5mp,silver': 1, 'akita': 1, 'camera,': 39, 'lenco': 14, '***olympus': 2, 'gateway': 3, 'transformers': 1, 'fe-5010': 2, 'micro': 2, 'apple': 1, 'xias': 3, 'carl': 1, 'takara': 4, 'pentacon': 5, 'ge-a735': 2, 'samsung,': 1, 'kaiser': 1, 'bratz': 1, 'i-carl

In [55]:
for x in sorted(list(map((lambda x : (first_word_not_manuf[x],x)), list(first_word_not_manuf)))):
    print(x)

(1, '.3mp')
(1, '.42x')
(1, '12.0')
(1, '12.0mp')
(1, '12mp')
(1, '2-way')
(1, '227593:')
(1, '228563:')
(1, '2600')
(1, '262905:')
(1, '27')
(1, '28')
(1, '2upc')
(1, '2upc-r205')
(1, '2upc-r71')
(1, '3')
(1, '3-in-1')
(1, '3.1mp')
(1, '3.2')
(1, '3.5"')
(1, '30')
(1, '310')
(1, '4.7w')
(1, '40')
(1, '48')
(1, '4gb')
(1, '5mp')
(1, '8gb')
(1, '8mp')
(1, '[mju:]')
(1, 'a835')
(1, 'ad-elegance')
(1, 'advent')
(1, 'akita')
(1, 'akor')
(1, 'ansmann')
(1, 'apc')
(1, 'apn')
(1, 'appareils-photos')
(1, 'apple')
(1, 'aqua')
(1, 'arctic')
(1, 'b-tech')
(1, 'baby')
(1, 'batterie')
(1, 'ben')
(1, 'biostek')
(1, 'black')
(1, 'bratz')
(1, 'bushnell')
(1, 'cadre\xa0photo\xa0/\xa0reveil\xa0/\xa0radi')
(1, 'cadre\xa0photo\xa0numerique\xa07')
(1, 'cam')
(1, 'camcorder,')
(1, 'camera,digital,5mp,silver')
(1, 'camera,digital,mini')
(1, 'caméra')
(1, 'car')
(1, 'carl')
(1, 'cb350')
(1, 'cell/ipod/gps/mp3/smartphone/digital')
(1, 'che-ez!')
(1, 'click')
(1, 'cokin')
(1, 'colorpix')
(1, 'compact')
(1, 'con

In [56]:
products_manu_model = {}
for prod in products:
    if prod['manufacturer'].lower() in products_manu_model:
        products_manu_model[prod['manufacturer'].lower()].append(prod['model'])
    else:
        products_manu_model[prod['manufacturer'].lower()] = [prod['model']]

In [60]:
print(products_manu_model)

{'pentax': ['WS80', 'E70', 'H90', 'X70', 'I-10', 'K-r', 'E80', 'EI-200', 'E85', 'K-5', 'RS1000', 'M85', 'P70', 'P80', 'M90', 'X90', 'S1', 'W90', 'K-x', 'WG-1 GPS', 'WG-1', 'K-7', 'W80'], 'fujifilm': ['T205', '1500', 'Z900EXR', 'AX305', 'A170', 'S2500HD', 'AX300', 'S1500', '4700 Zoom', 'A100', 'JX200', '4900 Zoom', '2400 Zoom', 'JZ500', 'AV250', 'S3400', 'S2990', '1200', 'AV205', 'Z33WP', 'J30', 'MX-2700', '1700 Zoom', 'Z808EXR', 'S4050', 'JX400', 'S1800', 'MX-700', 'AV255', '40i', 'S1 Pro', 'F305EXR', 'F70EXR', 'T300', 'AX355', 'AX350', 'S200EXR', 'F80EXR', 'S3350', 'S4000', 'T200', 'Z800EXR', 'AV200', 'Z35', 'JX300', 'MX-2900 Zoom', 'S2950', 'Z300', 'S3200', 'F550 EXR', 'S2600HD', 'JX305', 'JX350', 'XP10', 'DS-260HD', 'HS10', 'S3300', 'Z700EXR', '2900Z', 'MX-1200', 'JX205', 'F500 EXR', '1400 Zoom', 'HS20', 'S1600', 'Z30', '700', 'JX355', 'S2800HD', 'S1880', '1300', 'JZ505', 'DS-300', 'Z91', 'HS11', 'F200EXR', 'JX250', 'X100', 'HS22', 'S2900HD', 'MX-600 Zoom', 'S3450', 'T305', 'Z90', '

In [59]:
for entry in products_manu_model:
    print(entry, len(products_manu_model[entry]))

pentax 23
fujifilm 97
kyocera 2
leica 8
panasonic 80
sigma 4
olympus 94
kodak 51
agfa 7
contax 1
sanyo 4
samsung 70
konica minolta 5
toshiba 6
casio 43
sony 81
epson 11
ricoh 14
hp 7
canon 84
nikon 51


In [61]:
all_model = []
for entry in products_manu_model:
    all_model.extend(products_manu_model[entry])
print(sorted(all_model))

['100', '100 HS', '100 IS', '1000 HS', '105', '10S', '110 IS', '1100D', '115 HS', '120 IS', '1200', '1280', '130 IS', '1300', '1400 Zoom', '1500', '1680', '1700 Zoom', '200 IS', '200F', '210 IS', '220 HS', '2300', '2330', '2400 Zoom', '2700', '2900Z', '300', '300 HS', '300 HS', '3000 Zoom', '30S', '310 HS', '3300', '350', '4.3', '400F', '40i', '4700 Zoom', '4900 Zoom', '5', '500', '500 HS', '500D', '5010', '50S', '550', '550D', '550WP', '600', '600', '600', '600D', '60D', '650', '700', '700', '700', '7000', '7010', '7030', '7040', '750 Zoom', '7D', '800', '800', '850 Zoom', '880', '900', '9000', '900S', '9010', '95 IS', '950', '990', '990 IS', 'A100', 'A1100 IS', 'A1200', 'A150', 'A170', 'A2100 IS', 'A2200', 'A3000 IS', 'A3100 IS', 'A3200 IS', 'A3300 IS', 'A480', 'A490', 'A495', 'A5', 'A5 Zoom', 'A50', 'A800', 'AQ100', 'AV200', 'AV205', 'AV250', 'AV255', 'AX300', 'AX305', 'AX350', 'AX355', 'C-2000 Zoom', 'C-2020 Zoom', 'C-2040 Zoom', 'C-21', 'C-2100 UZ', 'C-2500 L', 'C-3000 Zoom', 'C-3

In [65]:
no_family = 0
for prod in products:
    if 'family' not in prod:
        no_family += 1
print(no_family, no_family/len(products))

258 0.3472409152086137


{'dfe': 11, 'dfse': 12}

KeyError: 'ddd'

In [70]:
# find +
contain_plus_sign = 0
for listing in listings:
    if listing['title'].find('&') != -1:
        print(listing['title'])
        contain_plus_sign += 1
print(contain_plus_sign)

Nikon EN-EL9a 1080mAh Ultra High Capacity Li-ion Battery Pack for Nikon D40, D40x, D60, D3000, & D5000 Digital SLR Cameras
Nikon ML-L3 Wireless Remote Control for Nikon P7000, D3000, D40, D40x, D50, D5000, D60, D70, D7000, D70s, D80 & D90 Digital SLR Cameras
NEW WP5300 Blue(with Mirco 2GB)Waterproof 12MP Digital Camera& Video Recorder
DURAGADGET Padded Camera Bag With Shoulder Strap & Zip Pockets For Go Pro Hero HD Head Cams (Helmet Hero, Motorsports Hero, Surf Hero)
DURAGADGET Utra-fine anti-static brush & lens pen cleaner for Kodak Playtouch, Playfull
DURAGADGET Utra-fine anti-static brush & lens pen cleaner for Canon Ixus 1000HS, Ixus 300HS, Ixus 210, Ixus 200IS, Ixus 130, Ixus 120IS
DURAGADGET Utra-fine anti-static brush & lens pen cleaner for Panasonic Lumix FH27, FH25, FP5, FP7, FH5, FH2, S3, S1
DURAGADGET Utra-fine anti-static brush & lens pen cleaner for Toshiba Camileo S30, S20
DURAGADGET Utra-fine anti-static brush & lens pen cleaner for Kodak Easyshare Mini M200, M583, M522,

In [67]:
# find "for"
contain_for = 0
for listing in listings:
    if listing['title'].find('for') != -1:
        contain_for += 1
print(contain_for)

496


In [72]:
# create a data structure to store result
result = {}
for prod in products:
    result[prod['manufacturer']] = [] # used to add listings

In [75]:
# create a data structure to store products : {manufacturer:(family,model,name)}
products_dict = {}
for prod in products:
    if prod['manufacturer'].lower() not in products_dict:
        products_dict[prod['manufacturer'].lower()] = []
    products_dict[prod['manufacturer'].lower()].append((prod['family'] if 'family' in prod else '', prod['model'], prod['product_name']))

In [76]:
products_dict

{'agfa': [('ePhoto', '1280', 'Agfa_ePhoto_1280'),
  ('ePhoto', 'CL30 Clik!', 'Agfa_ePhoto_CL30_Clik'),
  ('', 'DC-600uw', 'Agfa_DC-600uw'),
  ('ePhoto', 'CL30', 'Agfa_ePhoto_CL30'),
  ('ePhoto', 'CL18', 'Agfa_ePhoto_CL18'),
  ('ePhoto', '1680', 'Agfa_ePhoto_1680'),
  ('ePhoto', 'CL50', 'Agfa_ePhoto_CL50')],
 'canon': [('Digital IXUS', '130 IS', 'Canon_Digital_IXUS_130_IS'),
  ('PowerShot', 'SD980 IS', 'Canon_PowerShot_SD980_IS'),
  ('IXUS', '310 HS', 'Canon-IXUS-310HS'),
  ('Digital IXUS', '95 IS', 'Canon_Digital_IXUS_95_IS'),
  ('PowerShot', 'A5 Zoom', 'Canon_PowerShot_A5_Zoom'),
  ('PowerShot', 'A5', 'Canon_PowerShot_A5'),
  ('EOS', '60D', 'Canon_EOS_60D'),
  ('PowerShot', 'A3300 IS', 'Canon-A3300-IS'),
  ('Digital IXUS', '200 IS', 'Canon_Digital_IXUS_200_IS'),
  ('PowerShot', 'A3200 IS', 'Canon-A3200-IS'),
  ('Digital IXUS', '990 IS', 'Canon_Digital_IXUS_990_IS'),
  ('ELPH', '500 HS', 'Canon-ELPH-500HS'),
  ('PowerShot', 'G1', 'Canon_PowerShot_G1'),
  ('IXY', '200F', 'Canon_IXY_200F

# skip content after " for ", " with ", " w/ ", " + ",  and " & "
# Fuji == fujifilm && agfa == agfaphoto
# skip all color word
# skip content in brackets
# skip "the"

# track the first non-digit word => as manufacturer
## - skip (105, 'fototasche,') (138, 'polaroid') (140, 'easypix') (263, 'vivitar') (283, 'rollei') (288, 'ge') (291, 'duragadget')
## - 

# if manufacturer is not matched, full grade is 3
# else (manufacturer is matched), full grade is 2
#    if there is family name exists, family worth 1 point and model worth 1 point
#    else model worth 2 point

# assume n words contained in a family/model, then each matching of word worth 1/n points

# take the max value, if it is > (full grade - 1) (not inclusive), then take the argmax index; 
# otherwise leave it un-signed

In [78]:
products_dict['olympus']

[('', 'mju 9010', 'Olympus_mju_9010'),
 ('', 'T110', 'Olympus-T-110'),
 ('', 'C-2500 L', 'Olympus_C-2500_L'),
 ('', 'FE-5010', 'Olympus_FE-5010'),
 ('', 'C-2000 Zoom', 'Olympus_C-2000_Zoom'),
 ('', 'mju Tough 8010', 'Olympus_mju_Tough_8010'),
 ('', 'PEN E-P2', 'Olympus_PEN_E-P2'),
 ('', 'T100', 'Olympus-T100'),
 ('', 'VG110', 'Olympus-vg110'),
 ('Tough', 'TG-810', 'Olympus-TG-810'),
 ('', 'mju 7030', 'Olympus_mju_7030'),
 ('', 'mju Tough 6010', 'Olympus_mju_Tough_6010'),
 ('Stylus', '7010', 'Olympus_Stylus_7010'),
 ('', 'C-3040 Zoom', 'Olympus_C-3040_Zoom'),
 ('', 'VG120', 'Olympus-VG120'),
 ('', 'C-2100 UZ', 'Olympus_C-2100_UZ'),
 ('Stylus', '9000', 'Olympus_Stylus_9000'),
 ('', 'SP-610 UZ', 'Olympus-SP610UZ'),
 ('', 'C-3000 Zoom', 'Olympus_C-3000_Zoom'),
 ('', 'E-450', 'Olympus_E-450'),
 ('', 'PEN E-P1', 'Olympus_PEN_E-P1'),
 ('', 'FE-4040', 'Olympus_FE-4040'),
 ('Stylus', 'Tough-3000', 'Olympus_Stylus_Tough-3000'),
 ('', 'D-450 Zoom', 'Olympus_D-450_Zoom'),
 ('Stylus', '7000', 'Olym

In [34]:
# check the first word
first_word_not_manuf_count_re = 0
first_word_not_manuf_re = dict()
pattern = r'\b[^\d^\s^\W]+\b'
for listing in listings:
    try:
        first_word = re.search(pattern,listing['title']).group(0)
    except Exception as e:
        print(listing['title'])
    if first_word.lower() not in manu:
        first_word_not_manuf_count_re += 1
        if first_word.lower() in first_word_not_manuf_re:
            first_word_not_manuf_re[first_word.lower()] += 1
        else:
            first_word_not_manuf_re[first_word.lower()] = 1

2UPC-R710


In [35]:
first_word_not_manuf_re

{'a': 2,
 'acer': 8,
 'acmaxx': 3,
 'ad': 1,
 'advance': 5,
 'advanced': 5,
 'advent': 1,
 'afs': 1,
 'agfaphoto': 95,
 'aiptek': 7,
 'airis': 3,
 'akita': 1,
 'akor': 1,
 'alpha': 4,
 'ansmann': 1,
 'antquarium': 3,
 'apc': 1,
 'apn': 1,
 'appareil': 29,
 'appareils': 1,
 'apple': 1,
 'aqua': 1,
 'aquashot': 3,
 'architect': 2,
 'arctic': 1,
 'argus': 12,
 'ativa': 2,
 'b': 1,
 'baby': 1,
 'barbie': 4,
 'batterie': 1,
 'battery': 2,
 'bell': 16,
 'ben': 1,
 'benq': 32,
 'bilora': 5,
 'biostek': 1,
 'black': 4,
 'blue': 3,
 'bosch': 3,
 'bratz': 1,
 'braun': 4,
 'bundle': 12,
 'bushnell': 1,
 'cadre': 2,
 'cam': 1,
 'camcorder': 1,
 'camera': 45,
 'caméra': 1,
 'canal': 2,
 'caplio': 2,
 'car': 1,
 'carl': 1,
 'casecrown': 2,
 'cell': 1,
 'che': 1,
 'childs': 2,
 'click': 1,
 'cm': 1,
 'cnm': 3,
 'cobra': 18,
 'cokin': 1,
 'coleman': 5,
 'colorpix': 1,
 'combelle': 2,
 'compact': 2,
 'concord': 12,
 'cool': 1,
 'cooler': 1,
 'coolpix': 17,
 'cp': 1,
 'crayola': 8,
 'creative': 2,
 'cru